janitor/utils.py

```diff
@@ -2373,3 +2376,719 @@ def _sub_process_text_result_MultiIndex(index: pd.MultiIndex, result, df):
+    if strict:
+        right_c = right_c.take(positions)
+        left_c = left_c.repeat(search_indices)
+        rows_equal = left_c.array == right_c.array
+        if rows_equal.all():
+            return None
+        if rows_equal.any():
+            left_c = left_c.index[~rows_equal]
+            right_c = right_c.index[~rows_equal]
+            return left_c, right_c
+        return left_c.index, right_c.index
+
+    right_c = right_c.index.take(positions)
+    left_c = left_c.index.repeat(search_indices)
+
+    return left_c, right_c
+
+
+def _greater_than_indices(
+    left_c: pd.Series,
+    right_c: pd.Series,
+    len_conditions: int,
+    strict: bool,
+    not_equal: bool = False,
+):
+    """
+    Use binary search to get indices where left_c is greater than
+    or equal to right_c. If strict is True,
+    then only indices where `left_c`
+    is greater than (but not equal to) `right_c` are returned.
+
+    Returns a tuple of (left_c, right_c)
+    if len_conditions is == 1, else left_c.
+    """
+
+    # quick break, avoiding the hassle
+    if left_c.max() < right_c.min():
+        return None
+
+    if not_equal is False:
+        if right_c.hasnans:
+            right_c = right_c.dropna()
+        if not right_c.is_monotonic_increasing:
+            right_c = right_c.sort_values()
+    if left_c.hasnans:
+        left_c = left_c.dropna()
+
+    search_indices = right_c.searchsorted(left_c, side="right")
+    # if any of the positions in `search_indices`
+    # is equal to 0
+    # that means the respective position in `left_c`
+    # has no values from `right_c` that are greater than
+    # or equal, and should therefore be discarded
+    rows_equal = search_indices == 0
+    if rows_equal.any():
+        left_c = left_c[~rows_equal]
+        search_indices = search_indices[~rows_equal]
+    if search_indices.size == 0:
+        return None
+
+    if len_conditions > 1:
+        return left_c.index
+    # for each index in `search_indices`,
+    # generate all indices for `right_keys`,
+    # where the values in `right_keys` are less than
+    # or equal to `left_c`
+    positions = _ge_create_ranges(search_indices)
+    if strict:
+        right_c = right_c.take(positions)
+        left_c = left_c.repeat(search_indices)
+        rows_equal = left_c.array == right_c.array
+        if rows_equal.all():
+            return None
+        if rows_equal.any():
+            left_c = left_c.index[~rows_equal]
+            right_c = right_c.index[~rows_equal]
+            return left_c, right_c
+        return left_c.index, right_c.index
+
+    right_c = right_c.index.take(positions)
+    left_c = left_c.index.repeat(search_indices)
+
+    return left_c, right_c
+
+
+def _create_conditional_join_empty_frame(
+    df: pd.DataFrame, right: pd.DataFrame, how: str
+):
+    """
+    Create final dataframe for conditional join,
+    if there are no matches.
+    """
+
+    if how == "inner":
```

Comment:
Magic Strings can be changed to Constants / Enums and used across ensuring more type safety and consistency though I'm aware python type checking isn't exactly enforcing but just for guidance